include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1185,39 +1212,25 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
                     __dpl_sycl::__group_barrier(__item_id);
 
                     // 2. Find any element that satisfies pred and set local atomic value to global atomic
-                    constexpr auto __comp = typename _BrickTag::_Compare{};
-                    __pred(__item_id, __n_iter, __wgroup_size, __comp, __found_local, __brick_tag, __rngs...);
+                    __pred(__item_id, __n_iter, __wgroup_size, __found_local, __brick_tag, __rngs...);
                     __dpl_sycl::__group_barrier(__item_id);
 
                     // Set local atomic value to global atomic
-                    if (__local_idx == 0 && __comp(__found_local.load(), __found.load()))
+                    if (__local_idx == 0)
                     {
-                        if constexpr (__or_tag_check)
-                            __found.store(1);
-                        else
+                        const auto __found_local_state = __found_local.load();
+                        if (__found_local_state != __init_value)
                         {
-                            const auto __found_local_state = __found_local.load();
-
-                            bool __compare_exchange_processed = false;
-                            for (auto __old = __found.load();
-                                 !__compare_exchange_processed && __comp(__found_local_state, __old);
-                                 __old = __found.load())
-                            {
-                                // If we replace the atomic value successfully, we should break the loop to avoid extra operations with atomic
-                                __compare_exchange_processed =
-                                    __found.compare_exchange_strong(__old, __found_local_state);
-                            }
+                            // Update global (for all groups) atomic state with the found index
+                            _BrickTag::__save_state_to_atomic(__found, __found_local_state);
                         }
                     }
                 });
         });
         //The end of the scope  -  a point of synchronization (on temporary sycl buffer destruction)
     }
 
-    if constexpr (__or_tag_check)
-        return __result;
-    else
-        return __result != __init_value ? __result : __rng_n;
+    return _BrickTag::__eval_res(__result, __rng_n);
```

Comment:
This comes probably down to preference but I preferred the approach before. I think it was easier to read and didn't require duplicated routines in forward/backward tag.

Reply:
Let's resolve this point a little bit later.